File: scripts/local-index.py
#!/usr/bin/env python3
"""
로컬 인덱싱 파이프라인 - ChromaDB PersistentClient + Ollama
서버 없이 로컬에서 검색 인덱스 구축
"""

import argparse
import json
import time
import hashlib
import requests
from pathlib import Path
from typing import List, Dict, Any, Optional

# ChromaDB는 PersistentClient 사용 (서버 불필요)
import chromadb

# 설정
OLLAMA_URL = "http://localhost:11434"
EMBEDDING_MODEL = "nomic-embed-text"
COLLECTION_NAME = "search-index"
CHROMA_PATH = "./local-index/chroma-data"

# 시드 URL 데이터
SEED_URLS = {
    "tech-docs": [
        "https://developers.cloudflare.com/",
        "https://react.dev/",
        "https://vuejs.org/",
        "https://angular.io/",
        "https://svelte.dev/",
        "https://nextjs.org/",
        "https://nuxt.com/",
        "https://astro.build/",
        "https://remix.run/",
        "https://expressjs.com/",
        "https://fastify.dev/",
        "https://hono.dev/",
        "https://docs.docker.com/",
        "https://kubernetes.io/docs/",
        "https://terraform.io/docs/",
        "https://aws.amazon.com/docs/",
        "https://cloud.google.com/docs/",
        "https://learn.microsoft.com/",
        "https://developer.mozilla.org/",
        "https://typescriptlang.org/docs/",
        "https://graphql.org/learn/",
        "https://tailwindcss.com/",
        "https://vitejs.dev/",
        "https://vitest.dev/",
        "https://playwright.dev/",
        "https://cypress.io/",
        "https://jestjs.io/",
        "https://mochajs.org/",
        "https://eslint.org/",
        "https://prettier.io/",
        "https://webpack.js.org/",
        "https://rollupjs.org/",
        "https://esbuild.github.io/",
        "https://parceljs.org/",
    ],
    "programming": [
        "https://docs.python.org/3/",
        "https://doc.rust-lang.org/book/",
        "https://go.dev/doc/",
        "https://docs.oracle.com/en/java/",
        "https://developer.mozilla.org/en-US/docs/Web/JavaScript",
        "https://docs.swift.org/swift-book/",
        "https://kotlinlang.org/docs/",
        "https://dart.dev/",
        "https://elixir-lang.org/learn/",
        "https://www.haskell.org/tutorial/",
        "https://clojure.org/guides/getting_started",
        "https://scala-lang.org/documentation/",
        "https://learn.microsoft.com/dotnet/",
        "https://www.gnu.org/software/bash/manual/",
        "https://docs.gitlab.com/ee/",
        "https://docs.github.com/en",
    ],
    "news-intl": [
        "https://rss.nytimes.com/services/xml/rss/nyt/HomePage.xml",
        "https://feeds.bbci.co.uk/news/rss.xml",
        "https://www.theguardian.com/world/rss",
        "https://feeds.washingtonpost.com/rss/world",
        "https://www.aljazeera.com/xml/rss/all.xml",
        "https://rss.cnn.com/rss/edition.rss",
        "https://feeds.skynews.com/feeds/rss/world.xml",
        "https://www.france24.com/en/rss",
        "https://www.dw.com/en/rss/s-30978029",
    ],
    "news-kr": [
        "https://rss.naver.com/rss.xml",
        "https://www.chosun.com/arc/outboundfeeds/rss/",
        "https://www.joongang.co.kr/rss",
        "https://www.hankyung.com/rss/all-news",
        "https://www.mk.co.kr/rss/30100041/",
        "https://www.donga.com/news/RSS",
        "https://www.seoul.co.kr/rss/news.xml",
        "https://www.khan.co.kr/rss/rss_data/kh_latest.xml",
        "https://www.news1.kr/rss/",
    ],
    "news-jp": [
        "https://www.asahi.com/rss/",
        "https://www.yomiuri.co.jp/rss/",
        "https://www.nikkei.com/rss/",
        "https://mainichi.jp/rss/",
        "https://www.sankei.com/rss/",
        "https://www.tokyo-np.co.jp/rss/",
    ],
    "news-cn": [
        "https://www.sina.com.cn/rss/",
        "https://news.qq.com/rss/",
        "https://www.36kr.com/feed",
        "https://www.ifanr.com/feed",
        "https://sspai.com/feed",
        "https://www.huxiu.com/rss/",
        "https://www.chinanews.com.cn/rss/",
    ],
    "science": [
        "https://www.nature.com/nature.rss",
        "https://arxiv.org/rss/cs.AI",
        "https://arxiv.org/rss/cs.CL",
        "https://arxiv.org/rss/cs.LG",
        "https://arxiv.org/rss/cs.CV",
        "https://arxiv.org/rss/cs.SE",
    ],
}


class LocalIndexer:
    def __init__(self, chroma_path: str = CHROMA_PATH, ollama_url: str = OLLAMA_URL):
        self.ollama_url = ollama_url
        # PersistentClient 사용 (서버 불필요)
        self.client = chromadb.PersistentClient(path=chroma_path)
        self.collection = None

    def init_collection(self) -> bool:
        """ChromaDB 컬렉션 초기화"""
        try:
            self.collection = self.client.get_or_create_collection(
                name=COLLECTION_NAME,
                metadata={"hnsw:space": "cosine"},
            )
            print(f"✅ 컬렉션 초기화: {COLLECTION_NAME} (ID: {self.collection.id})")
            return True
        except Exception as e:
            print(f"❌ 컬렉션 초기화 실패: {e}")
            return False

    def get_embedding(self, text: str) -> List[float]:
        """Ollama로 임베딩 생성"""
        try:
            response = requests.post(
                f"{self.ollama_url}/api/embeddings",
                json={"model": EMBEDDING_MODEL, "prompt": text},
            )
            if response.status_code == 200:
                return response.json().get("embedding", [])
            else:
                print(f"⚠️ 임베딩 생성 실패: {response.text}")
                return []
        except Exception as e:
            print(f"⚠️ Ollama 연결 실패: {e}")
            return []

    def chunk_text(self, text: str, max_tokens: int = 300) -> List[str]:
        """텍스트 청킹 (단어 기반)"""
        words = text.split()
        chunks = []
        current_chunk = []

        for word in words:
            current_chunk.append(word)
            if len(current_chunk) >= max_tokens:
                chunks.append(" ".join(current_chunk))
                current_chunk = []

        if current_chunk:
            chunks.append(" ".join(current_chunk))

        return chunks if chunks else [text[:1000]]

    def index_url(self, url: str, title: str = "", content: str = "") -> bool:
        """단일 URL 인덱싱"""
        try:
            if not content:
                content = title or url

            chunks = self.chunk_text(content)

            for i, chunk in enumerate(chunks[:5]):  # 최대 5개 청크
                embedding = self.get_embedding(chunk)
                if not embedding:
                    continue

                doc_id = hashlib.md5(f"{url}_{i}".encode()).hexdigest()
                metadata = {
                    "url": url,
                    "title": title or url,
                    "chunk_index": i,
                    "domain": url.split("/")[2] if "/" in url else url,
                    "indexed_at": time.strftime("%Y-%m-%dT%H:%M:%SZ"),
                }

                self.collection.upsert(
                    ids=[doc_id],
                    embeddings=[embedding],
                    metadatas=[metadata],
                    documents=[chunk],
                )

            print(f"✅ 인덱싱 완료: {url} ({len(chunks)}개 청크)")
            return True

        except Exception as e:
            print(f"❌ 인덱싱 실패 ({url}): {e}")
            return False

    def index_urls(self, urls: List[str], category: str = "all") -> Dict[str, Any]:
        """여러 URL 인덱싱"""
        results = {"success": 0, "failed": 0, "total": len(urls)}

        for i, url in enumerate(urls):
            print(f"\n[{i+1}/{len(urls)}] 인덱싱 중: {url}")

            if self.index_url(url, title=url, content=url):
                results["success"] += 1
            else:
                results["failed"] += 1

            time.sleep(0.5)

        return results

    def get_stats(self) -> Dict[str, Any]:
        """인덱스 통계"""
        if self.collection:
            return {
                "collection": COLLECTION_NAME,
                "count": self.collection.count(),
            }
        return {"collection": COLLECTION_NAME, "count": 0}


def main():
    parser = argparse.ArgumentParser(description="로컬 인덱싱 파이프라인")
    parser.add_argument(
        "--category",
        default="tech-docs",
        choices=list(SEED_URLS.keys()) + ["all"],
        help="인덱싱할 카테고리",
    )
    parser.add_argument("--limit", type=int, default=10, help="인덱싱할 URL 수 제한")
    parser.add_argument("--urls", nargs="+", help="직접 URL 목록 지정")
    parser.add_argument("--stats", action="store_true", help="인덱스 통계 출력")
    parser.add_argument("--clear", action="store_true", help="인덱스 초기화")

    args = parser.parse_args()

    indexer = LocalIndexer()

    if args.clear:
        indexer.init_collection()
        indexer.client.delete_collection(COLLECTION_NAME)
        print(f"🗑️ 컬렉션 삭제 완료: {COLLECTION_NAME}")
        return

    if not indexer.init_collection():
        return

    if args.stats:
        stats = indexer.get_stats()
        print(f"\n📊 인덱스 통계:")
        print(f"   컬렉션: {stats['collection']}")
        print(f"   문서 수: {stats['count']}")
        return

    # URL 목록 준비
    if args.urls:
        urls = args.urls
    elif args.category == "all":
        urls = []
        for cat_urls in SEED_URLS.values():
            urls.extend(cat_urls)
    else:
        urls = SEED_URLS.get(args.category, [])

    urls = urls[: args.limit]

    print(f"\n🚀 로컬 인덱싱 시작")
    print(f"   카테고리: {args.category}")
    print(f"   URL 수: {len(urls)}")
    print(f"   ChromaDB: {CHROMA_PATH} (PersistentClient)")
    print(f"   Ollama: {OLLAMA_URL}")
    print(f"   임베딩 모델: {EMBEDDING_MODEL}")
    print()

    start_time = time.time()
    results = indexer.index_urls(urls, args.category)
    elapsed = time.time() - start_time

    print(f"\n{'='*60}")
    print(f"✅ 인덱싱 완료!")
    print(f"   성공: {results['success']}/{results['total']}")
    print(f"   실패: {results['failed']}/{results['total']}")
    print(f"   소요 시간: {elapsed:.1f}초")
    print(f"{'='*60}")

    stats = indexer.get_stats()
    print(f"\n📊 현재 인덱스: {stats['count']}개 문서")


if __name__ == "__main__":
    main()

File: scripts/local-index.ts
#!/usr/bin/env -S npx tsx
/**
 * 로컬 인덱싱 스크립트
 *
 * ChromaDB + Ollama를 사용한 완전 로컬 인덱싱
 * Cloudflare 없이 로컬에서 인덱싱/검색 가능
 *
 * 사용법:
 *   # 서비스 시작
 *   docker run -p 8000:8000 chromadb/chroma
 *   ollama serve
 *   ollama pull nomic-embed-text
 *
 *   # 인덱싱 실행
 *   npx tsx scripts/local-index.ts --urls urls.txt
 *   npx tsx scripts/local-index.ts --category=tech
 *   npx tsx scripts/local-index.ts --search "react hooks"
 */

import * as fs from 'fs'
import { LocalIndexingService, type LocalDocument } from '../src/lib/local-indexing'

// ============================================================
// Seed URLs by Category
// ============================================================

const TECH_DOCS = [
  'https://react.dev/learn',
  'https://react.dev/reference/react',
  'https://vuejs.org/guide/introduction.html',
  'https://nextjs.org/docs',
  'https://nuxt.com/docs',
  'https://www.typescriptlang.org/docs/',
  'https://docs.python.org/3/',
  'https://fastapi.tiangolo.com/',
  'https://nodejs.org/docs/latest/api/',
  'https://go.dev/doc/',
  'https://doc.rust-lang.org/book/',
  'https://docs.docker.com/',
  'https://kubernetes.io/docs/',
  'https://developers.cloudflare.com/workers/',
  'https://developers.cloudflare.com/d1/',
  'https://developers.cloudflare.com/vectorize/',
  'https://docs.aws.amazon.com/',
  'https://www.postgresql.org/docs/',
  'https://docs.mongodb.com/',
  'https://redis.io/docs/',
  'https://docs.pytorch.org/',
  'https://www.tensorflow.org/guide',
  'https://huggingface.co/docs/transformers/',
  'https://docs.github.com/',
  'https://git-scm.com/doc',
]

const NEWS_SITES = [
  'https://www.bbc.com/news',
  'https://www.nytimes.com/',
  'https://www.theguardian.com/',
  'https://www.cnn.com/',
  'https://www.reuters.com/',
  'https://techcrunch.com/',
  'https://www.theverge.com/',
  'https://arstechnica.com/',
  'https://www.wired.com/',
  'https://www.cnbc.com/',
]

const WIKI_DOCS = [
  'https://en.wikipedia.org/wiki/Machine_learning',
  'https://en.wikipedia.org/wiki/Artificial_intelligence',
  'https://en.wikipedia.org/wiki/Deep_learning',
  'https://en.wikipedia.org/wiki/Natural_language_processing',
  'https://en.wikipedia.org/wiki/Computer_vision',
  'https://en.wikipedia.org/wiki/Neural_network',
  'https://en.wikipedia.org/wiki/Transformer_(deep_learning)',
  'https://en.wikipedia.org/wiki/Large_language_model',
  'https://en.wikipedia.org/wiki/Cloud_computing',
  'https://en.wikipedia.org/wiki/Internet_of_things',
]

// ============================================================
// Main Script
// ============================================================

interface LocalIndexArgs {
  mode: 'index' | 'search' | 'stats'
  urls?: string[]
  category?: string
  query?: string
  topK?: number
  chromaUrl?: string
  ollamaUrl?: string
}

/**
 * Exit with a failure code. Unlike `process.exit` (typed `any` by
 * @cloudflare/workers-types, so it does not terminate control flow), the
 * explicit `never` return type lets tsc narrow variables after guard clauses.
 */
function fail(code: number): never {
  process.exit(code)
  throw new Error('unreachable')
}

function parseArgs(): LocalIndexArgs {
  const args = process.argv.slice(2)
  const options: LocalIndexArgs = {
    mode: 'index',
    topK: 10,
    chromaUrl: 'http://localhost:8000',
    ollamaUrl: 'http://localhost:11434',
  }

  for (const arg of args) {
    if (arg.startsWith('--urls=')) {
      const filePath = arg.split('=')[1]
      options.urls = fs
        .readFileSync(filePath, 'utf-8')
        .split('\n')
        .map((l) => l.trim())
        .filter((l) => l.length > 0 && l.startsWith('http'))
    } else if (arg.startsWith('--category=')) {
      options.category = arg.split('=')[1]
    } else if (arg.startsWith('--search=')) {
      options.mode = 'search'
      options.query = arg.split('=').slice(1).join('=')
    } else if (arg === '--stats') {
      options.mode = 'stats'
    } else if (arg.startsWith('--top-k=')) {
      options.topK = parseInt(arg.split('=')[1], 10)
    } else if (arg.startsWith('--chroma=')) {
      options.chromaUrl = arg.split('=')[1]
    } else if (arg.startsWith('--ollama=')) {
      options.ollamaUrl = arg.split('=')[1]
    }
  }

  return options
}

async function fetchDocument(url: string): Promise<LocalDocument | null> {
  try {
    const response = await fetch(url, {
      headers: { 'User-Agent': 'LocalIndexer/1.0' },
      signal: AbortSignal.timeout(10000),
    })

    if (!response.ok) return null

    const html = await response.text()

    // 간단한 HTML 파싱
    const titleMatch = html.match(/<title[^>]*>([\s\S]*?)<\/title>/i)
    const title = titleMatch ? titleMatch[1].replace(/<[^>]+>/g, '').trim() : url

    // 본문 추출 (간단한 휴리스틱)
    const content = html
      .replace(/<script[\s\S]*?<\/script>/gi, '')
      .replace(/<style[\s\S]*?<\/style>/gi, '')
      .replace(/<[^>]+>/g, ' ')
      .replace(/\s+/g, ' ')
      .trim()
      .slice(0, 10000) // 최대 10,000자

    const domain = new URL(url).hostname.replace(/^www\./, '')

    return {
      id: Buffer.from(url).toString('base64').slice(0, 32),
      url,
      title,
      content,
      domain,
    }
  } catch (err) {
    console.error(`문서 가져오기 실패: ${url}`, err)
    return null
  }
}

async function runIndexing(args: LocalIndexArgs) {
  console.log('═══════════════════════════════════════════════════════')
  console.log('  로컬 인덱싱 스크립트')
  console.log('═══════════════════════════════════════════════════════')

  // URL 목록 준비
  let urls: string[] = []
  if (args.urls) {
    urls = args.urls
  } else if (args.category) {
    switch (args.category) {
      case 'tech':
        urls = TECH_DOCS
        break
      case 'news':
        urls = NEWS_SITES
        break
      case 'wiki':
        urls = WIKI_DOCS
        break
      case 'all':
        urls = [...TECH_DOCS, ...NEWS_SITES, ...WIKI_DOCS]
        break
      default:
        console.error(`알 수 없는 카테고리: ${args.category}`)
        process.exit(1)
    }
  } else {
    urls = [...TECH_DOCS, ...NEWS_SITES, ...WIKI_DOCS]
  }

  console.log(`  URL 수: ${urls.length}`)
  console.log(`  ChromaDB: ${args.chromaUrl}`)
  console.log(`  Ollama: ${args.ollamaUrl}`)
  console.log('')

  // 서비스 초기화
  const service = new LocalIndexingService({
    chromaUrl: args.chromaUrl,
    ollamaUrl: args.ollamaUrl,
  })

  const status = await service.initialize()
  if (!status.chroma || !status.ollama) {
    console.error('❌ 서비스 미준비:')
    console.error(`  ChromaDB: ${status.chroma ? '✅' : '❌'}`)
    console.error(`  Ollama: ${status.ollama ? '✅' : '❌'}`)
    console.error('')
    console.error('시작 방법:')
    console.error('  docker run -p 8000:8000 chromadb/chroma')
    console.error('  ollama serve')
    console.error('  ollama pull nomic-embed-text')
    process.exit(1)
  }

  console.log('✅ 서비스 준비 완료')
  console.log('')

  // 인덱싱 실행
  let successCount = 0
  let failCount = 0

  for (let i = 0; i < urls.length; i++) {
    const url = urls[i]
    const progress = `[${i + 1}/${urls.length}]`

    process.stdout.write(`${progress} ${url}... `)

    const doc = await fetchDocument(url)
    if (!doc) {
      console.log('❌ 실패')
      failCount++
      continue
    }

    const result = await service.indexDocument(doc)
    if (result) {
      console.log('✅ 성공')
      successCount++
    } else {
      console.log('❌ 실패')
      failCount++
    }
  }

  console.log('')
  console.log('═══════════════════════════════════════════════════════')
  console.log(`  ✅ 완료: ${successCount} 성공, ${failCount} 실패`)
  console.log('═══════════════════════════════════════════════════════')
}

async function runSearch(args: LocalIndexArgs) {
  if (!args.query) {
    console.error('검색어를 입력하세요: --search="react hooks"')
    fail(1)
  }
  const query = args.query

  console.log('═══════════════════════════════════════════════════════')
  console.log('  로컬 검색')
  console.log('═══════════════════════════════════════════════════════')
  console.log(`  쿼리: ${query}`)
  console.log(`  Top-K: ${args.topK}`)
  console.log('')

  const service = new LocalIndexingService({
    chromaUrl: args.chromaUrl,
    ollamaUrl: args.ollamaUrl,
  })

  await service.initialize()

  const results = await service.search(query, args.topK)

  if (results.length === 0) {
    console.log('검색 결과 없음')
    return
  }

  console.log(`검색 결과: ${results.length}건`)
  console.log('')

  for (let i = 0; i < results.length; i++) {
    const r = results[i]
    console.log(`${i + 1}. ${r.title}`)
    console.log(`   URL: ${r.url}`)
    console.log(`   스코어: ${r.score.toFixed(4)}`)
    console.log('')
  }
}

async function runStats(args: LocalIndexArgs) {
  console.log('═══════════════════════════════════════════════════════')
  console.log('  로컬 인덱스 상태')
  console.log('═══════════════════════════════════════════════════════')

  const service = new LocalIndexingService({
    chromaUrl: args.chromaUrl,
    ollamaUrl: args.ollamaUrl,
  })

  const _status = await service.initialize()
  const stats = await service.getStats()

  console.log(`  ChromaDB: ${stats.chromaReady ? '✅ 연결됨' : '❌ 미연결'}`)
  console.log(`  Ollama: ${stats.ollamaReady ? '✅ 연결됨' : '❌ 미연결'}`)
  console.log(`  총 문서: ${stats.totalDocuments}`)
  console.log(`  총 청크: ${stats.totalChunks}`)
  console.log('')
}

// ============================================================
// Run
// ============================================================

async function main() {
  const args = parseArgs()

  switch (args.mode) {
    case 'index':
      await runIndexing(args)
      break
    case 'search':
      await runSearch(args)
      break
    case 'stats':
      await runStats(args)
      break
  }
}

main().catch(console.error)
